### DISCUSSION TITLE
S2 Styles Question

### DISCUSSION THREAD

**Discussion by ninavizz** [adobe/react-spectrum#7398] [General]
Hello! Been poking around the S2 Storybook, and I noticed that many of the UI components lack an "Emphasis" option—where things can be blue'd, as opposed to remaining black and white. Notably, the Action Buttons, tabs, and segmented menus lack this option. Are there plans to add-in the "Emphasis" option before launch to parity with S1?

Also, following S2's release, how long will S1 remain supported for—and will migrating some components to S2, require migrating all components to S2?

**Comment by reidbarber** [comment]
Some component variants that are available in S1 might not be available in S2 (due to design or accessibility reasons). I don't want to speak for specific components since some component designs are still in progress and things could change.

Hopefully the [migration docs](https://react-spectrum.adobe.com/s2/index.html?path=/docs/migrating--docs) make it easier to upgrade and understand the changes. We are definitely open to feedback on that experience!

S2 will definitely be able to be used alongside S1. Although it isn't ideal, we recognize that teams may have complex apps that take a while to fully upgrade. I'm not sure of a support timeline for S1 but it definitely isn't going away in the foreseeable future.

**Comment by ninavizz** [comment]
@reidbarber Thx for the thorough response! Yes, the app I'm working on is very complicated—and its unique needs may require keeping S1 components—less because of time to migrate, and more because some items (such as the TableView) appear to be going out in S2 with far fewer features, than in S1. Likewise, to the "Emphasis" capability on buttons that I'm not seeing in S2 components. As the designer, I don't see aesthetic/usability conflicts with how I'm envisioning it—but it will be a pain for the devs. Thx again!

**Reply by snowystinger** [reply]
S2 hasn't had its v1 release yet, we're still working on bringing it up to feature parity where it makes sense, see [S2 feature parity with S1 planned?](https://github.com/adobe/react-spectrum/discussions/7433#top) in some cases components have more features than s1 had. So feature parity is a bit of a tricky thing to define.

"Emphasis" is less relied on, or it has moved to another component. For instance, ActionButton was split into a new ToggleButton which *does* carry the emphasized prop https://react-spectrum.adobe.com/s2/index.html?path=/docs/togglebutton--docs
This should be used instead, and it looks like maybe we missed it in the migration docs. @reidbarber was this one left out on purpose? or did we just not have enough information to handle it for people?

Tabs have had emphasized removed, I believe the common case where that was used is now handled by SegmentedControl which has a very different visual attached to it and emphasized doesn't make sense there.

I'm not sure what Segmented Menu is?

Hope that helps.

**Reply by ninavizz** [reply]
@snowystinger Hey Robert, thx for the reply! A "Segmented Menu" is the more common language I've heard, for what y'all are calling a SegmentedControl—so, we're talkinga bout the same thing. 

The "Emphasis" blue on both the Buttons and the Tabs is pretty important for how I've integrated them into what I'm using Spectrum with—so, oyy. I also noticed that on the Toggle button you linked to, the Emphasis Blue in S2 appears to be a hue of Indigo, and not the Blue900 currently being used?

I'm using the SegmentedControl from S2 already (!!) as I found it much more elegant than the Action Buttons glom/hack from S1; but I still need to keep the Emphasis Blue on the Horizontal and Vertical tabs from S1. 

Thx for the more detailed answer on S2's release; yes, feature parity is something I've been looking at a lot, and can appreciate how hard it is to hit as a moving target.

**Reply by snowystinger** [reply]
Possibly what you're looking for in terms of Tabs is actually going to be our SideNav, which we haven't implemented yet, but should in the nearish future. I can't advise more since I don't know how you're using the component.

Otherwise, S2 is a major version update, of the Spectrum Design system as a whole. Some concepts are going to be discarded in light of feedback or research, colors may change, new features added, etc.

We do have ways for you to create your own design system that aren't affected by changes in the Spectrum system. React Aria Components are completely unstyled and are how we've built our implementation of S2 so quickly. We also now have a theming system (sort of like PandaCSS/Tailwind/etc) which you could use as an inspiration/starting point to quickly style your components. You can also refer to our source for how we've implemented our S2 version of the components.

I hope this unblocks you.

**Reply by ninavizz** [reply]
@snowystinger I'm not at all blocked, don't worry! Your taking the time to share all of this information with me so that I can plan and set-up my PM for success, is incredibly helpful. 

The "Tabs" component (its name, I believe?) from S1's Web library has both a Horizontal and a Vertical option. My understanding from another thread, is that the horizontal option is being (or has already been?) depreciated; so using a styled Aria component for that, I believe is the running plan with the team I'm supporting.

Is the new Theming system something we could download from NPM and put to use with Aria components—or is it only available with the full S2 version update? I also love that in S2 the use of the "new" color system (that S1 introduced), seems to be influenced by WCAG 2.0's luminescence standard vs the 1.0 "color contrast" standard (that regressed usability for sighted users).

I've advised my client in the risks of using Spectrum as a pre-built system optimized for another company's needs, driven by opaque-to-us factors. TY again, for all you've shared! :)

**Reply by snowystinger** [reply]
Tabs S2 has both horizontal and vertical, see https://react-spectrum.adobe.com/s2/?path=/docs/tabs--docs

S2 specific theme-ing is currently only available if S2 is installed. However, the underlying technology for it can be used regardless and the theme itself could be copied and modified.
https://react-spectrum.adobe.com/s2/?path=/docs/style-macro--docs

**Reply by ninavizz** [reply]
The prob with S2's tabs is that they lack the Emphasis distinction (making the selected object bright blue); which, for usability, is an important attribute in the data-dense context of its use in the enterprise app I'm working on. Hence, looking to Aria as a fallback. 

So, the Style Macro (which looks RAD!) is an option to apply to a hybrid Aria/S1 install—as a sort of "theme" for it? Nope, not seeking to provide user-facing "themes" as an option for end users; but would love to make it easier for devs to implement some consistent type/color mods to existing styles to enhance usability/a11y.

**Reply by snowystinger** [reply]
> they lack the Emphasis distinction

This is why I thought maybe SideNav might be what you're looking for, but I can't really show you what that looks like as it doesn't exist yet.

StyleMacro doesn't have anything to do with S1, and it isn't a user facing theme for end users. It's for our developers to easily apply S2 styles (colors, spacing, transitions, etc) to RAC for more one-off designs that are specific to their product.

My suggestion was that the theme definition could be modified to be whatever your team needs. For example, you could copy the theme and alter the colors if you can't use the new button emphasized colors. The S2 components will always have our S2 theme baked into them, so you wouldn't affect those.

We haven't exported any of that as API or tooling yet, possibly one day, but again, the underlying tech already exists.

**Reply by ninavizz** [reply]
@snowystinger Ok—If I'm semi-understanding you (and I cannot understand Javascript to save my life, so that's not saying much!)... I think you're telling me that a developer with a solid understanding of Javascript and possibly some of React, could use the "StyleMacro" from S1 and apply it to Aria, with any modifications we want—and from a developer perspective, that'd function as a custom theme within the code itself?

(TY again for your patience & clarity, you rule!)

**Reply by dannify** [reply]
Sort of. The style macro is a way to apply styles defined by Spectrum (in this case it is styles and colors for Spectrum 2) and apply those to unstyled components. React Aria Components are the same components we use for Spectrum with the Spectrum styles added to them. If there is a component that does not work for you style wise, you are better off having your devs take the component you need and add the custom styles you need. You could even look at the code we use for our S2 components to use as the starting point or you could use the many examples we have on the website.

**Reply by cheema-corellian** [reply]
As I understand it, the new style macro makes it easier to customize styling of existing S2 components.  Something easy like this where the marginStart style value is allowed to be changed.

```
<S2ActionButton styles={style({marginStart: 8})}>
  Edit
</S2ActionButton>
```

Using this method, changing many CSS properties is supported.  However changing colors is not.  For that, you have to use the underlying React Aria component and customize it.  The task then goes from a 3-line thing to a 500-line thing.  As seen here:

https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/s2/src/ActionButton.tsx

Above link shows the code for S2 ActionButton that builds on top of React Aria ActionButton.  Code for some dependencies is not shown on that page, but would have to be included.

Granted, a dev would not have to write the entire 500-line component from scratch.  They could just copy the S2 ActionButton and edit it to the get the desired colors.  But, at that point they would have forked the component and are largely responsible for its care and feeding from there on out.

If the above is a correct understanding, then perhaps the use of UNSAFE_style and UNSAFE_classNames with S2 components becomes quite appealing as they bring the task back down to a 3-line thing.  The risks associated with the use of these facilities become quite palatable, given the alternative.

Please let me know if I got any of the details wrong and if there is an easier/better way to customize the colors of S2 components.

**Reply by devongovett** [reply]
Customizing the styles of spectrum components is not a supported use case. As a design system, Spectrum is about bringing consistency to our products, and allowing things to be overridden defeats the purpose of that. If you want to build your own custom design system, then I'd recommend using React Aria Components. UNSAFE should almost never be used (it is actually unsafe and will break when we make updates).

**Reply by cheema-corellian** [reply]
@devongovett I completely agree with what you said about Spectrum goals and that is one of the many reasons why we use Spectrum and not Aria.  I am just thinking out loud here about options available to a team should they find themselves in a situation where they need to change a single color value on an S2 component.   And also trying to validate my assumptions.
